# 5. Tá»‘i Æ°u Performance (Optimization)

## ğŸ§  useMemo - Memoize Expensive Calculations

**âœ… KHI NÃ€O DÃ™NG:**

1. **Filter/map/reduce array lá»›n (> 50 items)**
```javascript
const filteredPlaces = useMemo(() => {
  return places.filter(place => {
    const matchDistrict = !filters.district || place.district === filters.district;
    const matchPrice = place.priceRange.max <= filters.maxPrice;
    const matchMood = filters.mood ? place.aiTags.mood.includes(filters.mood) : true;
    return matchDistrict && matchPrice && matchMood;
  });
}, [places, filters.district, filters.maxPrice, filters.mood]);
```

2. **Transform data tá»« API**
```javascript
const placesByDistrict = useMemo(() => {
  return places.reduce((acc, place) => {
    if (!acc[place.district]) acc[place.district] = [];
    acc[place.district].push(place);
    return acc;
  }, {});
}, [places]);
```

3. **TÃ­nh toÃ¡n phá»©c táº¡p (> 5ms)**
```javascript
const averageRating = useMemo(() => {
  if (!reviews.length) return 0;
  const sum = reviews.reduce((acc, r) => acc + r.rating, 0);
  return (sum / reviews.length).toFixed(1);
}, [reviews]);
```

**âŒ KHI KHÃ”NG NÃŠN DÃ™NG:**

```javascript
// âŒ BAD: TÃ­nh toÃ¡n quÃ¡ Ä‘Æ¡n giáº£n
const fullName = useMemo(() => {
  return `${firstName} ${lastName}`; // < 0.1ms, khÃ´ng cáº§n memo
}, [firstName, lastName]);

// âŒ BAD: Táº¡o object nhá»
const style = useMemo(() => ({ color: 'red' }), []); // KhÃ´ng cáº§n memo
```

---

## ğŸ¯ useCallback - Memoize Functions

**âœ… KHI NÃ€O DÃ™NG:**

1. **Function Ä‘Æ°á»£c pass xuá»‘ng child component (trÃ¡nh re-render)**
```javascript
const PlaceList = ({ places }) => {
  const handlePlaceSelect = useCallback((placeId) => {
    navigate(`/places/${placeId}`);
    trackEvent('place_clicked', { placeId });
  }, [navigate]);

  return places.map(place => (
    <PlaceCard 
      key={place._id}
      place={place}
      onSelect={handlePlaceSelect} // KhÃ´ng táº¡o function má»›i má»—i render
    />
  ));
};
```

2. **Function lÃ  dependency cá»§a useEffect/useMemo**
```javascript
const fetchPlaceDetails = useCallback(async (placeId) => {
  const data = await placeService.getPlaceById(placeId);
  setPlaceData(data);
}, []);

useEffect(() => {
  fetchPlaceDetails(placeId);
}, [placeId, fetchPlaceDetails]); // KhÃ´ng trigger re-fetch vÃ´ Ã­ch
```

3. **Event handlers trong lists**
```javascript
const handleReviewSubmit = useCallback((reviewData) => {
  submitReview.mutate({ placeId, ...reviewData });
}, [placeId, submitReview]);
```

**âŒ KHI KHÃ”NG NÃŠN DÃ™NG:**

```javascript
// âŒ BAD: Function chá»‰ dÃ¹ng local, khÃ´ng pass xuá»‘ng
const handleClick = useCallback(() => {
  console.log('clicked'); // KhÃ´ng cáº§n memo
}, []);

// âŒ BAD: Event handler inline Ä‘Æ¡n giáº£n
<button onClick={useCallback(() => setCount(c => c + 1), [])}>
  // QuÃ¡ phá»©c táº¡p cho viá»‡c Ä‘Æ¡n giáº£n
</button>
```

---

## ğŸ›¡ï¸ React.memo - Prevent Component Re-renders

**âœ… KHI NÃ€O DÃ™NG:**

1. **List items (PlaceCard, ReviewCard)**
```javascript
const PlaceCard = React.memo(({ place, onSelect }) => {
  return (
    <div className="place-card" onClick={() => onSelect(place._id)}>
      <img src={place.images[0]} alt={place.name} />
      <h3>{place.name}</h3>
      <p>{formatPrice(place.priceRange.min)}</p>
    </div>
  );
});

PlaceCard.displayName = 'PlaceCard';
```

2. **Pure UI components (Button, Badge, Icon)**
```javascript
const Button = React.memo(({ children, onClick, variant = 'primary' }) => {
  return (
    <button className={`btn btn-${variant}`} onClick={onClick}>
      {children}
    </button>
  );
});
```

3. **Heavy components (Charts, Maps)**
```javascript
const PlaceMap = React.memo(({ places, center }) => {
  return <MapView places={places} center={center} />;
}, (prevProps, nextProps) => {
  // Custom comparison: chá»‰ re-render khi places hoáº·c center thay Ä‘á»•i
  return prevProps.places.length === nextProps.places.length &&
         prevProps.center.lat === nextProps.center.lat;
});
```

**âŒ KHI KHÃ”NG NÃŠN DÃ™NG:**

```javascript
// âŒ BAD: Component props thay Ä‘á»•i liÃªn tá»¥c
const Counter = React.memo(({ count }) => {
  return <div>{count}</div>; // count thay Ä‘á»•i má»—i giÃ¢y, memo vÃ´ Ã­ch
});

// âŒ BAD: Component quÃ¡ Ä‘Æ¡n giáº£n
const Text = React.memo(({ children }) => <p>{children}</p>); // KhÃ´ng cáº§n
```

---

## ğŸš« Anti-Patterns Pháº£i TrÃ¡nh

```javascript
// âŒ BAD: Inline object trong props â†’ táº¡o má»›i má»—i render
<PlaceCard style={{ margin: 10 }} />

// âœ… GOOD: Hoist ra ngoÃ i
const cardStyle = { margin: 10 };
<PlaceCard style={cardStyle} />

// âŒ BAD: Inline array trong props
<PlaceCard tags={['cafe', 'quiet']} />

// âœ… GOOD: useMemo hoáº·c constant
const tags = useMemo(() => ['cafe', 'quiet'], []);
<PlaceCard tags={tags} />

// âŒ BAD: Anonymous function trong prop
<button onClick={() => handleClick(id)}>Click</button>

// âœ… GOOD: useCallback
const onClick = useCallback(() => handleClick(id), [id]);
<button onClick={onClick}>Click</button>

// âŒ BAD: Nested map/filter trong render
{places.map(p => p.reviews.filter(r => r.rating > 4).map(...))}

// âœ… GOOD: useMemo
const topReviews = useMemo(() => {
  return places.flatMap(p => 
    p.reviews.filter(r => r.rating > 4)
  );
}, [places]);
```

---

## ğŸ“Š Optimization Decision Tree

```
CÃ³ pháº£i tÃ­nh toÃ¡n phá»©c táº¡p (> 5ms)?
â”œâ”€ YES â†’ useMemo
â””â”€ NO â†’ KhÃ´ng cáº§n optimize

Function Ä‘Æ°á»£c pass xuá»‘ng child?
â”œâ”€ YES â†’ useCallback
â””â”€ NO â†’ KhÃ´ng cáº§n optimize

Component re-render khÃ´ng cáº§n thiáº¿t?
â”œâ”€ YES â†’ React.memo
â””â”€ NO â†’ KhÃ´ng cáº§n optimize
```

---

## ğŸ” Profiling & Measurement

**Tools:**

- **React DevTools Profiler**: Record vÃ  phÃ¢n tÃ­ch render time
- **Chrome DevTools Performance**: Flame chart toÃ n bá»™ app
- **Lighthouse**: Check Core Web Vitals

**Metrics má»¥c tiÃªu:**

- First Contentful Paint (FCP) < 1.8s
- Largest Contentful Paint (LCP) < 2.5s
- Total Blocking Time (TBT) < 300ms
- Cumulative Layout Shift (CLS) < 0.1

**Khi nÃ o cáº§n optimize:**

1. Component render > 16ms (60 FPS)
2. User interaction bá»‹ lag
3. Lighthouse score < 90
