# 6. React Query Best Practices

## üîë Query Keys Structure (Hierarchical)

```javascript
// ‚úÖ GOOD: T·ªï ch·ª©c query keys theo hierarchy
['places'] // Base key
['places', 'list'] // All places
['places', 'list', { district: 'Ba Dinh' }] // Filtered places
['places', 'detail', placeId] // Single place
['places', 'reviews', placeId] // Place's reviews

['chat', 'history', userId] // Chat history
['chat', 'message', messageId] // Single message

['user', 'profile'] // Current user
['user', 'preferences'] // User preferences
```

**L·ª£i √≠ch:**

- D·ªÖ invalidate queries: `invalidateQueries(['places'])` ‚Üí clear t·∫•t c·∫£ place-related
- Consistent v√† predictable

---

## üíæ Caching Strategy

```javascript
// hooks/usePlaces.js
export const usePlaces = (filters = {}) => {
  return useQuery({
    queryKey: ['places', 'list', filters],
    queryFn: () => placeService.getPlaces(filters),
    
    // Caching config
    staleTime: 5 * 60 * 1000, // 5 ph√∫t - data c√≤n "fresh"
    gcTime: 10 * 60 * 1000, // 10 ph√∫t - gi·ªØ trong cache
    
    // Refetch behavior
    refetchOnWindowFocus: false, // Kh√¥ng refetch khi switch tab
    refetchOnReconnect: true, // Refetch khi reconnect internet
    
    // Retry
    retry: 2, // Retry 2 l·∫ßn khi fail
    retryDelay: (attemptIndex) => Math.min(1000 * 2 ** attemptIndex, 30000),
  });
};
```

**Caching rules theo data type:**


| Data Type | staleTime | gcTime | L√Ω do |
| :-- | :-- | :-- | :-- |
| Static data (categories, districts) | 30 mins | 1 hour | √çt thay ƒë·ªïi |
| User profile | 10 mins | 30 mins | Thay ƒë·ªïi trung b√¨nh |
| Place details | 5 mins | 15 mins | C√≥ th·ªÉ update |
| Search results | 2 mins | 5 mins | Thay ƒë·ªïi th∆∞·ªùng xuy√™n |
| Chat messages | 0 | 2 mins | Real-time data |

---

## üîÑ Optimistic Updates (UX T·ªët h∆°n)

```javascript
// hooks/useReviews.js
export const useSubmitReview = (placeId) => {
  const queryClient = useQueryClient();
  
  return useMutation({
    mutationFn: (reviewData) => reviewService.submitReview(placeId, reviewData),
    
    // OPTIMISTIC UPDATE
    onMutate: async (newReview) => {
      // 1. Cancel outgoing refetches (tr√°nh race condition)
      await queryClient.cancelQueries(['places', 'reviews', placeId]);
      
      // 2. Snapshot current data (ƒë·ªÉ rollback n·∫øu fail)
      const previousReviews = queryClient.getQueryData(['places', 'reviews', placeId]);
      
      // 3. Optimistically update UI (user th·∫•y ngay)
      queryClient.setQueryData(['places', 'reviews', placeId], (old) => {
        return [...(old || []), { ...newReview, _id: 'temp-id', createdAt: new Date() }];
      });
      
      return { previousReviews };
    },
    
    onError: (err, newReview, context) => {
      // 4. Rollback n·∫øu API fail
      queryClient.setQueryData(
        ['places', 'reviews', placeId],
        context.previousReviews
      );
      toast.error('Kh√¥ng th·ªÉ g·ª≠i ƒë√°nh gi√°');
    },
    
    onSuccess: () => {
      toast.success('ƒê√°nh gi√° ƒë√£ ƒë∆∞·ª£c g·ª≠i');
    },
    
    onSettled: () => {
      // 5. Refetch ƒë·ªÉ sync v·ªõi server
      queryClient.invalidateQueries(['places', 'reviews', placeId]);
      queryClient.invalidateQueries(['places', 'detail', placeId]); // Update avg rating
    },
  });
};
```

---

## üì° Prefetching (TƒÉng Performance)

```javascript
// Prefetch place details khi hover PlaceCard
const PlaceCard = ({ place }) => {
  const queryClient = useQueryClient();
  
  const handleMouseEnter = () => {
    queryClient.prefetchQuery({
      queryKey: ['places', 'detail', place._id],
      queryFn: () => placeService.getPlaceById(place._id),
      staleTime: 60000, // Cache 1 ph√∫t
    });
  };
  
  return (
    <div onMouseEnter={handleMouseEnter}>
      <Link to={`/places/${place._id}`}>{place.name}</Link>
    </div>
  );
};
```

---

## üîÅ Pagination v·ªõi keepPreviousData

```javascript
export const usePlaces = (page, filters) => {
  return useQuery({
    queryKey: ['places', 'list', page, filters],
    queryFn: () => placeService.getPlaces({ page, ...filters }),
    keepPreviousData: true, // Gi·ªØ data c≈© khi chuy·ªÉn trang ‚Üí kh√¥ng blink
  });
};

// Usage trong component
const PlacesPage = () => {
  const [page, setPage] = useState(1);
  const { data, isLoading, isFetching } = usePlaces(page, filters);
  
  return (
    <>
      <PlaceList places={data?.places} />
      {isFetching && <LoadingOverlay />} // Show loading khi fetch page m·ªõi
      <Pagination page={page} onPageChange={setPage} />
    </>
  );
};
```
